src/usb.c

```diff
@@ -0,0 +1,1619 @@
+
+      g_variant_builder_add (&filtered_devices_builder,
+                             "(s@a{sv}@a{sv})",
+                             device_id,
+                             g_steal_pointer (&device_variant),
+                             g_variant_dict_end (&device_options_dict));
+    }
+
+  *out_filtered_devices = g_variant_builder_end (&filtered_devices_builder);
+  return TRUE;
+}
+
+static XdpOptionKey usb_access_devices_options[] = {
+};
+
+static gboolean
+handle_acquire_devices (XdpDbusUsb            *object,
+                        GDBusMethodInvocation *invocation,
+                        const char            *arg_parent_window,
+                        GVariant              *arg_devices,
+                        GVariant              *arg_options)
+{
+  g_autoptr(XdpDbusImplRequest) impl_request = NULL;
+  g_autoptr(UsbSenderInfo) sender_info = NULL;
+  g_autoptr(GVariant) filtered_devices = NULL;
+  g_autoptr(GVariant) options = NULL;
+  g_autoptr(GError) error = NULL;
+  GVariantBuilder options_builder;
+  Permission permission;
+  Request *request;
+  Usb *self;
+
+  self = (Usb *) object;
+  request = request_from_invocation (invocation);
+
+  g_debug ("[usb] Handling AccessDevices");
+
+  REQUEST_AUTOLOCK (request);
+
+  permission = get_permission_sync (xdp_app_info_get_id (request->app_info),
+                                    PERMISSION_TABLE,
+                                    PERMISSION_ID);
+  if (permission == PERMISSION_NO)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                             "Not allowed to create USB sessions");
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  impl_request = xdp_dbus_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (usb_impl)),
+                                                       G_DBUS_PROXY_FLAGS_NONE,
+                                                       g_dbus_proxy_get_name (G_DBUS_PROXY (usb_impl)),
+                                                       request->id,
+                                                       NULL,
+                                                       &error);
+  if (!impl_request)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options,
+                           &options_builder,
+                           usb_access_devices_options,
+                           G_N_ELEMENTS (usb_access_devices_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+  options = g_variant_builder_end (&options_builder);
```

Comment:
options is an autoptr but `g_variant_builder_end` returns a floating reference which is not a proper reference and calling g_object_unref via the autoptr is an error.

I see this pattern everywhere in x-d-p though and I'm not sure if I'm missing something. /cc @GeorgesStavracas

Reply:
I have added a sink in this specific location.